rfcs/text/0000-analytics-instrumentation-plan.md

```diff
@@ -0,0 +1,137 @@
+
+* `category` - the name of the service issuing the event, e.g. `data-storage-service`, `data-portal`, or `dcp-cli`
+* `action` - is a human-readable string representing the function performed, e.g. `download-bundle` or `upload-file`
+* `label` - is the identifier of the entity acted on, e.g, bundle UUID or `<organ>-<method>-<data-contributor-last-name>`*
+
+### Frontend instrumentation
+
+For frontend components, we can use the standard GA instrumentation. Steps are included below.
+
+1. Front-end developers create trackers using the standard JavaScript tracking snippet    ([see
+   documentation](https://developers.google.com/analytics/devguides/collection/gtagjs/)).
+2. Tracking IDs should be chosen based on its deployment environment.
+3. At this time, we suggest not customizing the tracking behavior in any way. We expect the default settings will be
+   sufficient for our needs, and we can iterate if they are not after we’ve captured early user data.
+
+### Metrics Exporter
+
+The metrics exporter is an application for exporting request/response information from server-side logs to Google
+Measurement Protocol API in batch. Its components are as follows.
+
+* A [Kinesis Data Firehose Stream to
+  s3](https://docs.aws.amazon.com/firehose/latest/dev/create-destination.html#create-destination-s3) (KDFS)
+* A lambda that listens to new s3 events generated by Kinesis Data Firehose, loads the request information and submits
+  it   to the Google Measurement Protocol API   [in
+  batch](https://developers.google.com/analytics/devguides/collection/protocol/v1/devguide#batch)
+
+### Data storage service instrumentation
+
+The Data Storage Service (DSS) should fire an event whenever data is requested and pushed to users. Steps are included
+below.
+
+* Upon each `GET /v1/bundles/:bundle_uuid` and `GET /v1/files/:file_uuid` request, a JSON log line is emitted to
+  `stdout` with the key-value pairs `{"log-msg-type": "analytics", "system": "data-storage-service", ...}` and the
+  request info
+* A subscription for the Metrics Exporter's KDFS is created on DSS's Cloudwatch Logs log group
+* The subscription filter limits the stream only to log entries with the key-value pairs outlined above
+
+Approach benefits
+
+* Scalable
+* Minimal performance overhead on DSS
+* Bulk operations reduce computation, cost, and API rate limiting
+* This instrumentation design can be repeated for other data consumer APIs
+
+Approach detractors
+
+* Not as simple as a direct API call
+* Not as timely as a direct API call
+* Who will manage access to the Google Analytics account with this data?
+* How will this data be publicly shared with the wider scientific community, if at all?
+
+### Unresolved questions
+
+* What is the identifier design in event schema?
+* Will analysts need extra tools to make sense of event details?
+
+### Drawbacks and Limitations
+
+* Deriving human-readable identifiers will be tricky to balance with keeping systems schema-agnostic.
+
+
```

Comment:
There are no statements in this RFC about how this proposal will be GDPR compliant. There needs to be some sort of statement about the consideration around this issue as user tracking creates GDPR concerns

Reply:
This is a good observation. I will address this thoroughly to make sure we get it right. Updates soon.

Reply:
I like that you will address GDPR compliance but I am not sure how we will do this when there are legal teams associated with the HCA (institutions making the DCP) currently evaluating what we need to do to be compliant. I would imagine only general assurances can be made here until we get more details from them.

Reply:
I think this RFC can propose the design issues where GDPR compliance is a consideration and how it can be addressed.

Final approval of the associated privacy policy and details of what information is stored and for how long will need to go past the legal team but I don't think they need to be involved at this stage.

Reply:
I do not think the RFC or the work in environments (excluding production) should be held back for legal review but before we move this implementation to production we will need to make sure legal has reviewed a written privacy statement and we have confirmed the implementation meets that statement.

Reply:
The current (approved) Data Portal GDPR statement (at DCP home page/privacy) does state "We collect the following personal data from users ... IP address ... Statistics on web pages visited" which is intended to cover usage of Google Analytics, etc. Perhaps this needs expanding.

Reply:
Update: This has been reviewed and approved by the HCA GDPR Task Force. I have added a Data Privacy section detailing the measures we are taking to be compliant.